Sources/StreamChat/Utils/CoreDataLazy.swift

```diff
@@ -0,0 +1,65 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+/// Makes it possible to lazily evaluate properties of `NSManagedObject`s.
+///
+/// Works like a "classic" lazy wrapper. The only different is that the `computeValue` closure is evaluated on the queue
+/// of the provided `NSManagedObjectContext`.
+///
+@propertyWrapper
+class CoreDataLazy<T> {
+    /// When the cached value is reset, this closure is used to lazily get a new value which is then cached again.
+    var computeValue: (() -> T)!
+    
+    /// The context on whose queue is the `computeValue` closure evaluated.
+    weak var context: NSManagedObjectContext?
+    
+    var wrappedValue: T {
+        var returnValue: T!
+
+        // We need to make the changes inside the `mutate` block to ensure the wrapper is thread-safe.
+        __cached.mutate { value in
+            
+            if let value = value {
+                returnValue = value
+                return
+            }
+
+            let perform = {
+                log.assert(self.computeValue != nil, "You must set the `computeValue` closure before accessing the cached value.")
+                returnValue = self.computeValue()
+            }
+            
+            if let context = context {
+                context.performAndWait { perform() }
+            } else {
+                // This is really a fallback for cases like tests, mocks, and edge cases. Even though `NSManagedObject.managedObjectContext`
+                // is optional, we probably won't reach this point during a normal production run.
```

Comment:
If we consider this to be unreachable in production, it might be better to add some assertion?

Reply:
@olejnjak I didn't say it's "unreachable" I said "we probably won't reach" 😅 The comment is actually outdated. MessageDTO uses this for the cases where the results are empty and we don't need to use the context queue. I'll update it.